src/services/blocks/trace/Trace.ts

```diff
@@ -0,0 +1,666 @@
+	// the root.
+	for (const span of spans) {
+		if (!span.parentId) {
+			continue;
+		}
+
+		if (!spansById.has(span.parentId)) {
+			throw new InternalServerError(
+				'Expected spans parent to exist in spansById'
+			);
+		}
+
+		spansById.get(span.parentId)?.children.push(span.id);
+	}
+
+	return { spansById, parsedSpans, executeBlockSpanId };
+}
+
+/**
+ * Find the Ids of al the descendant spans of `root`.
+ *
+ * @param root span which we want all the descendants of
+ * @param spansById map of span id => span with children
+ */
+function findDescendants(
+	root: number,
+	spansById: Map<number, SpanWithChildren>
+): number[] {
+	// Note: traversal order doesn't matter here
+	const stack = spansById.get(root)?.children;
+	if (!stack) {
+		return [];
+	}
+
+	const descendants = [];
+	while (stack.length) {
+		const curId = stack.pop();
+		if (!curId) {
+			break;
+		}
+
+		descendants.push(curId);
+
+		const curSpanChildren = spansById.get(curId)?.children;
+		curSpanChildren && stack.push(...curSpanChildren);
+	}
+
+	return descendants;
+}
+
+/**
+ * Extract extrinsic indexes from all `:extrinsic_index` storage item read events.
+ *
+ * @param spanIds List of spanIds for an apply extrinsic action group
+ * @param extrinsicIndexBySpanId Map of spanId => extrinsicIndex
+ */
+function extractExtrinsicIndex(
+	spanIds: number[],
+	extrinsicIndexBySpanId: Map<number, BN>
+): BN {
+	// There are typically multiple reads of extrinsic index when applying an extrinsic,
+	// So we get all of those reads, which should all be the same number.
+	const extrinsicIndex = spanIds.reduce((acc, id) => {
+		if (!extrinsicIndexBySpanId.has(id)) {
+			return acc;
+		}
+
+		acc.push(extrinsicIndexBySpanId.get(id) as BN);
+		return acc;
+	}, [] as BN[]);
+
+	if (extrinsicIndex.length === 0) {
+		// Since we assume the `spanIds` are for span from an apply extrinsic action group,
+		// we assume there is always an `:extrinsic_index` event
+		throw new InternalServerError('Expected at least one extrinsic index');
+	}
+
+	if (extrinsicIndex.length > 1) {
+		// If there are multiple reads we check that they are all the same. If they
+		// are this program has some incorrect assumptions and this error should
+		// be reported to the maintainers.
+		for (const [i, extIdx] of extrinsicIndex.entries()) {
+			if (i > 0 && !extrinsicIndex[i]?.eq(extIdx)) {
+				throw new InternalServerError(
+					'Expect extrinsic to only be applied at a single index.'
+				);
+			}
+		}
+	}
+
+	return extrinsicIndex[0];
+}
+
+/**
+ * Class for processing traces from the `state_traceBlock` RPC endpoint.
+ *
+ * Assumptions:
+ * - Spans do not start in sorted order.
+ * - Events start in sorted order.
+ *
+ * For a conceptual overview on traces from block excution and `Operation` construction
+ * consult [this diagram.](https://docs.google.com/drawings/d/1vZoJo9jaXlz0LmrdTOgHck9_1LsfuQPRmTr-5g1tOis/edit?usp=sharing)
```

Comment:
👍

Reply:
it would be super-useful add this in wiki/documentation in the long run.